src/diff/mod.rs

```diff
@@ -1,36 +1,285 @@
+
+        let mut old_iter = old.iter();
+        let mut new_iter = new.iter();
+        let mut old_entry_opt = old_iter.next();
+        let mut new_entry_opt = new_iter.next();
+
+        while old_entry_opt.is_some() || new_entry_opt.is_some() {
+            match (old_entry_opt, new_entry_opt) {
+                (Some(old_entry), Some(new_entry)) => {
+                    let cmp = new_entry.label().cmp(&old_entry.label());
+                    match cmp {
+                        Ordering::Greater => {
+                            diff.add_deleted_files(old_entry, parent_path)?;
+                            old_entry_opt = old_iter.next();
+                        },
+                        Ordering::Less => {
+                            diff.add_created_files(new_entry, parent_path)?;
+                            new_entry_opt = new_iter.next();
+                        },
+                        Ordering::Equal => {
+                            use DirectoryContents::{File, SubDirectory};
+                            match (new_entry, old_entry) {
+                                (File(new_file), File(old_file)) => {
+                                    if old_file.size != new_file.size || &old_file.checksum() != &new_file.checksum() {
+                                        diff.add_modified_file(new_file, parent_path);
+                                    }
+                                    old_entry_opt = old_iter.next();
+                                    new_entry_opt = new_iter.next();
+                                },
+                                (File(new_file), SubDirectory(old_dir)) => {
+                                    diff.add_created_file(new_file, parent_path);
+                                    diff.add_deleted_files(old_entry, parent_path)?;
+                                    old_entry_opt = old_iter.next();
+                                    new_entry_opt = new_iter.next();
+                                },
+                                (SubDirectory(new_dir), File(old_file)) => {
+                                    diff.add_created_files(new_entry, parent_path)?;
+                                    diff.add_deleted_file(old_file, parent_path);
+                                    old_entry_opt = old_iter.next();
+                                    new_entry_opt = new_iter.next();
+                                },
+                                (SubDirectory(new_dir), SubDirectory(old_dir)) => {
+                                    parent_path.borrow_mut().push(new_dir.label.clone());
+                                    Diff::collect_diff(&**old_dir, &**new_dir, parent_path, diff)?;
+                                    parent_path.borrow_mut().pop();
+                                    old_entry_opt = old_iter.next();
+                                    new_entry_opt = new_iter.next();
+                                },
+                                (_, _) => {
```

Comment:
I'd rather have an explicit match here and we just call `next` and let the loop do its thing. Realistically, we'll usually only see one `Repo` and that's at the root.

Reply:
Added explicit matches for Repo combinations.